pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/EtcdMetadataStore.java

```diff
@@ -0,0 +1,426 @@
+                                // Check that key does not exist
+                                txn.If(new Cmp(key, Cmp.Op.EQUAL, CmpTarget.createRevision(0)));
+                            } else {
+                                txn.If(new Cmp(key, Cmp.Op.EQUAL, CmpTarget.version(expectedVersion + 1)));
+                            }
+                        }
+                        break;
+                    }
+                    case DELETE: {
+                        OpDelete del = op.asDelete();
+                        ByteSequence key = ByteSequence.from(del.getPath(), StandardCharsets.UTF_8);
+                        if (del.getOptExpectedVersion().isPresent()) {
+                            txn.If(new Cmp(key, Cmp.Op.EQUAL,
+                                    CmpTarget.version(del.getOptExpectedVersion().get() + 1)));
+                        }
+                        break;
+                    }
+                }
+            });
+
+            // Then the requests
+            ops.forEach(op -> {
+                switch (op.getType()) {
+                    case GET: {
+                        txn.Then(
+                                Op.get(ByteSequence.from(op.asGet().getPath(), StandardCharsets.UTF_8),
+                                        SINGLE_GET_OPTION));
+                        break;
+                    }
+                    case PUT: {
+                        OpPut put = op.asPut();
+                        ByteSequence key = ByteSequence.from(put.getPath(), StandardCharsets.UTF_8);
+                        if (!put.getFuture().isDone()) {
+                            PutOption.Builder b = PutOption.newBuilder()
+                                    .withPrevKV();
+
+                            if (put.isEphemeral()) {
+                                b.withLeaseId(leaseId);
+                            }
+
+                            txn.Then(Op.put(key, ByteSequence.from(put.getData()), b.build()));
+                        }
+                        break;
+                    }
+                    case DELETE: {
+                        OpDelete del = op.asDelete();
+                        ByteSequence key = ByteSequence.from(del.getPath(), StandardCharsets.UTF_8);
+                        txn.Then(Op.delete(key, DeleteOption.DEFAULT));
+                        break;
+                    }
+                    case GET_CHILDREN: {
+                        OpGetChildren opGetChildren = op.asGetChildren();
+                        String path = opGetChildren.getPath();
+
+                        ByteSequence firstKey =
+                                ByteSequence.from(path.equals("/") ? path : path + "/",
+                                        StandardCharsets.UTF_8);
+                        ByteSequence lastKey = ByteSequence.from(path.equals("/") ? "0" : path + "0",
+                                StandardCharsets.UTF_8); // '0' is lexicographically just after '/'
```

Comment:
It's better to add more description here

```
If the range_end is one bit larger than the given key, then the range requests get the all keys with the prefix (the given key)
```

From the jetcd documentation, or use `withPrefix` directly which is easy to understand.

Reply:
Good point, fixed it with `withPrefix()`